rust/plugin-lib/src/base_cache.rs

```diff
@@ -27,49 +30,270 @@ const CHUNK_SIZE: usize = 1024 * 1024;
+            let resp = source.get_data(chunk_end, TextUnit::Utf8,
+                                       CHUNK_SIZE, self.rev)?;
+            self.append_chunk(resp);
+        }
+    }
+
+    /// Returns the offset of the line at `line_num`, zero-indexed, fetching
+    /// data from `source` if needed.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if `line_num` is greater than the total number of lines
+    /// in the document, or if there is a problem communicating with `source`.
+    pub fn offset_of_line<DS>(&mut self, source: &DS, line_num: usize) -> Result<usize, Error>
+        where DS: DataSource
+    {
+        if line_num > self.num_lines { return Err(Error::BadRequest) }
+        match self.cached_offset_of_line(line_num) {
+            Some(offset) => Ok(offset + self.offset),
+            None => {
+                let resp = source.get_data(line_num, TextUnit::Line, CHUNK_SIZE, self.rev)?;
+                self.reset_chunk(resp);
+                self.offset_of_line(source, line_num)
             }
         }
     }
 
+    /// Returns the offset of the provided `line_num` in `self.contents` if
+    /// it is present in the chunk.
+    fn cached_offset_of_line(&self, line_num: usize) -> Option<usize> {
+        if line_num < self.first_line { return None }
+
+        let rel_line_num = line_num - self.first_line;
+
+        if rel_line_num == 0 && self.first_line_offset == 0 {
+            return Some(0)
+        }
+        if rel_line_num <= self.line_offsets.len() {
+            return Some(self.line_offsets[rel_line_num - 1])
+        }
+
+        // EOF
+        if line_num == self.num_lines && self.offset + self.contents.len() == self.buf_size {
+            return Some(self.contents.len())
+        }
+        None
+    }
+
+    /// Clears anything in the cache up to `offset`, which is indexed relative
+    /// to `self.contents`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `offset` is not a character boundary, or if `offset` is greater than
+    /// the length of `self.content`.
+    fn clear_up_to(&mut self, offset: usize) {
+        if offset > self.contents.len() {
+            panic!("offset greater than content length: {} > {}", offset, self.contents.len())
+        }
+
+        let new_contents = self.contents.split_off(offset);
+        self.contents = new_contents;
+        self.offset += offset;
+        // first find out if offset is a line offset, and set first_line / first_line_offset
+        let (new_line, new_line_off) = match self.line_offsets.binary_search(&offset) {
+            Ok(idx) => (self.first_line + idx + 1, 0),
+            Err(0) => (self.first_line, self.first_line_offset + offset),
+            Err(idx) => (self.first_line + idx, offset - self.line_offsets[idx - 1]),
+        };
+
+        // then clear line_offsets up to and including offset
+        self.line_offsets = self.line_offsets.iter()
+            .filter(|i| **i > offset)
```

Comment:
Odd that this requires two `*`'s and the following none, but maybe I just don't understand the types...

Reply:
Indeed. As far as I understand it, the issue is mostly that `usize - &size` is okay but `usize > &usize` isn't.